File: VoiceRecorder.Audio/PitchShifter.cs
// this class based on code from awesomebox, a project created by by Ravi Parikh and Keegan Poppen, used with permission
// http://decabear.com/awesomebox.html
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Diagnostics;

namespace VoiceRecorder.Audio
{
    class PitchShifter
    {
        int detectedNote;
        protected float detectedPitch;  //set == inputPitch when shiftPitch is called
        protected float shiftedPitch;   //set == the target pitch when shiftPitch is called
        int numshifts;  //number of stored detectedPitch, shiftedPitch pairs stored for the viewer (more = slower, less = faster)
        Queue<PitchShift> shifts;
        protected int currPitch;
        protected int attack;
        int numElapsed;
        protected double vibRate;
        protected double vibDepth;
        double g_time;

        protected AutoTuneSettings settings;

        public PitchShifter(AutoTuneSettings settings)
        {
            this.settings = settings;
            numshifts = 5000;
            shifts = new Queue<PitchShift>(numshifts);

            currPitch = 0;
            attack = 0;
            numElapsed = 0;
            vibRate = 4.0;
            vibDepth = 0.00;
            g_time = 0.0;
        }

        protected float snapFactor(float freq)
        {
            float previousFrequency = 0.0f;
            float correctedFrequency = 0.0f;
            int previousNote = 0;
            int correctedNote = 0;
            for (int i = 1; i < 120; i++)
            {
                bool endLoop = false;
                foreach (int note in this.settings.AutoPitches)
                {
                    if (i % 12 == note)
                    {
                        previousFrequency = correctedFrequency;
                        previousNote = correctedNote;
                        correctedFrequency = (float)(8.175 * Math.Pow(1.05946309, (float)i));
                        correctedNote = i;
                        if (correctedFrequency > freq) { endLoop = true; }
                        break;
                    }
                }
                if (endLoop)
                {
                    break;
                }
            }
            if (correctedFrequency == 0.0) { return 1.0f; }
            int destinationNote = 0;
            double destinationFrequency = 0.0;
            // decide whether we are shifting up or down
            if (correctedFrequency - freq > freq - previousFrequency)
            {
                destinationNote = previousNote;
                destinationFrequency = previousFrequency;
            }
            else
            {
                destinationNote = correctedNote;
                destinationFrequency = correctedFrequency;
            }
            if (destinationNote != currPitch)
            {
                numElapsed = 0;
                currPitch = destinationNote;
            }
            if (attack > numElapsed)
            {
                double n = (destinationFrequency - freq) / attack * numElapsed;
                destinationFrequency = freq + n;
            }
            numElapsed++;
            return (float)(destinationFrequency / freq);
        }

        protected void updateShifts(float detected, float shifted, int targetNote)
        {
            if (shifts.Count >= numshifts) shifts.Dequeue();
            PitchShift shift = new PitchShift(detected, shifted, targetNote);
            Debug.WriteLine(shift);
            shifts.Enqueue(shift);
        }

        void setDetectedNote(float pitch)
        {
            for (int i = 0; i < 120; i++)
            {
                float d = (float)(8.175 * Math.Pow(1.05946309, (float)i) - pitch);
                if (-1.0 < d && d < 1.0)
                {
                    detectedNote = i;
                    return;
                }
            }
            detectedNote = -1;
        }

        bool isDetectedNote(int note)
        {
            return (note % 12) == (detectedNote % 12) && detectedNote >= 0;
        }

        protected float addVibrato(int nFrames)
        {
            g_time += nFrames;
            float d = (float)(Math.Sin(2 * 3.14159265358979 * vibRate * g_time / 44100) * vibDepth);
            return d;
        }
    }

    class SmbPitchShifter : PitchShifter
    {
        public SmbPitchShifter(AutoTuneSettings settings) : base(settings) { }

        void pitchShift(float factor, int nFrames, float[] buff)
        {
            //before the second nFrames was def_buffer_size, but def_buffer_size == nFrames (I think)
            // MRH: was nFrames but this is not a power of 2
            // 2048 works, let's try 1024
            int fftFrameSize = 2048;
            int osamp = 32; // 32 is best quality
            SmbPitchShift.smbPitchShift(factor, nFrames, fftFrameSize, osamp, 44100f, buff, buff);
        }

        public void ShiftPitch(float[] inputBuff, float inputPitch, float targetPitch, float[] outputBuff, int nFrames)
        {
            UpdateSettings();
            detectedPitch = inputPitch;
            float shiftFactor = 1.0f;
            if (this.settings.SnapMode)
            {
                if (inputPitch > 0)
                {
                    shiftFactor = snapFactor(inputPitch);
                    shiftFactor += addVibrato(nFrames);
                }
                if (shiftFactor > 2.0) shiftFactor = 2.0f;
                if (shiftFactor < 0.5) shiftFactor = 0.5f;

                float[] tempBuff = new float[nFrames];
                for (int i = 0; i < nFrames; i++)
                {
                    tempBuff[i] = (float)(inputBuff[i]);
                }

                pitchShift(shiftFactor, nFrames, tempBuff);

                for (int i = 0; i < nFrames; i++)
                {
                    outputBuff[i] = tempBuff[i]; // MRH: overwrite, don't accumulate
                }
            }
            else
            {
                //foreach (float midiPitch in pitches)
                float midiPitch = targetPitch;
                {
                    shiftFactor = 1.0f;
                    if (inputPitch > 0 && midiPitch > 0)
                    {
                        shiftFactor = midiPitch / inputPitch;
                    }

                    if (shiftFactor > 2.0) shiftFactor = 2.0f;
                    if (shiftFactor < 0.5) shiftFactor = 0.5f;

                    float[] tempBuff = new float[nFrames];
                    for (int i = 0; i < nFrames; i++)
                    {
                        tempBuff[i] = inputBuff[i];
                    }

                    pitchShift(shiftFactor, nFrames, tempBuff);

                    for (int i = 0; i < nFrames; i++)
                    {
                        outputBuff[i] = tempBuff[i];
                    }
                    //break;  //this line is a hack, because we are not polyphonic right now
                }
            }

            //vibrato
            //addVibrato(outputBuff, nFrames);

            shiftedPitch = inputPitch * shiftFactor;
            updateShifts(detectedPitch, shiftedPitch, this.currPitch);
        }

        private void UpdateSettings()
        {
            //these are going here, because this gets called once per frame
            vibRate = this.settings.VibratoRate;
            vibDepth = this.settings.VibratoDepth;
            attack = (int)((this.settings.AttackTimeMilliseconds * 441) / 1024.0);
        }
    }
    
    class PitchShift
    {
        public PitchShift(float detected, float shifted, int destNote)
        {
            this.DetectedPitch = detected;
            this.ShiftedPitch = shifted;
            this.DestinationNote = destNote;
        }

        public float DetectedPitch { get; private set; }
        public float ShiftedPitch { get; private set; }
        public int DestinationNote { get; private set; }

        public override string ToString()
        {
            return String.Format("detected {0:f2}Hz, shifted to {1:f2}Hz, {2}{3} ", DetectedPitch, ShiftedPitch,
                (Notes)(DestinationNote % 12),DestinationNote/12);
        }
    }
}

File: ~VB Port/VoiceRecorder.Audio/PitchShifter.vb
' this class based on code from awesomebox, a project created by by Ravi Parikh and Keegan Poppen, used with permission
' http://decabear.com/awesomebox.html
Imports System.Text

Namespace VoiceRecorder.Audio
    Friend Class PitchShifter
        Private detectedNote As Integer
        Protected detectedPitch As Single 'set == inputPitch when shiftPitch is called
        Protected shiftedPitch As Single 'set == the target pitch when shiftPitch is called
        Private numshifts As Integer 'number of stored detectedPitch, shiftedPitch pairs stored for the viewer (more = slower, less = faster)
        Private shifts As Queue(Of PitchShift)
        Protected currPitch As Integer
        Protected attack As Integer
        Private numElapsed As Integer
        Protected vibRate As Double
        Protected vibDepth As Double
        Private g_time As Double
        Protected sampleRate As Single

        Protected settings As AutoTuneSettings

        Public Sub New(ByVal settings As AutoTuneSettings, ByVal sampleRate As Single)
            Me.settings = settings
            Me.sampleRate = sampleRate
            numshifts = 5000
            shifts = New Queue(Of PitchShift)(numshifts)

            currPitch = 0
            attack = 0
            numElapsed = 0
            vibRate = 4.0
            vibDepth = 0.00
            g_time = 0.0
        End Sub

        Protected Function snapFactor(ByVal freq As Single) As Single
            Dim previousFrequency = 0.0F
            Dim correctedFrequency = 0.0F
            Dim previousNote = 0
            Dim correctedNote = 0
            For i = 1 To 119
                Dim endLoop = False
                For Each note As Integer In Me.settings.AutoPitches
                    If i Mod 12 = note Then
                        previousFrequency = correctedFrequency
                        previousNote = correctedNote
                        correctedFrequency = CSng(8.175 * Math.Pow(1.05946309, CSng(i)))
                        correctedNote = i
                        If correctedFrequency > freq Then
                            endLoop = True
                        End If
                        Exit For
                    End If
                Next note
                If endLoop Then
                    Exit For
                End If
            Next i
            If correctedFrequency = 0.0 Then
                Return 1.0f
            End If
            Dim destinationNote = 0
            Dim destinationFrequency = 0.0
            ' decide whether we are shifting up or down
            If correctedFrequency - freq > freq - previousFrequency Then
                destinationNote = previousNote
                destinationFrequency = previousFrequency
            Else
                destinationNote = correctedNote
                destinationFrequency = correctedFrequency
            End If
            If destinationNote <> currPitch Then
                numElapsed = 0
                currPitch = destinationNote
            End If
            If attack > numElapsed Then
                Dim n = (destinationFrequency - freq) / attack * numElapsed
                destinationFrequency = freq + n
            End If
            numElapsed += 1
            Return CSng(destinationFrequency / freq)
        End Function

        Protected Sub updateShifts(ByVal detected As Single, ByVal shifted As Single, ByVal targetNote As Integer)
            If shifts.Count >= numshifts Then
                shifts.Dequeue()
            End If
            Dim shift As New PitchShift(detected, shifted, targetNote)
            Debug.WriteLine(shift)
            shifts.Enqueue(shift)
        End Sub

        Private Sub setDetectedNote(ByVal pitch As Single)
            For i = 0 To 119
                Dim d = CSng(8.175 * Math.Pow(1.05946309, CSng(i)) - pitch)
                If -1.0 < d AndAlso d < 1.0 Then
                    detectedNote = i
                    Return
                End If
            Next i
            detectedNote = -1
        End Sub

        Private Function isDetectedNote(ByVal note As Integer) As Boolean
            Return (note Mod 12) = (detectedNote Mod 12) AndAlso detectedNote >= 0
        End Function

        Protected Function addVibrato(ByVal nFrames As Integer) As Single
            g_time += nFrames
            Dim d = CSng(Math.Sin(2 * 3.14159265358979 * vibRate * g_time / sampleRate) * vibDepth)
            Return d
        End Function
    End Class


    Friend Class SmbPitchShifter
        Inherits PitchShifter
        Public Sub New(ByVal settings As AutoTuneSettings, ByVal sampleRate As Single)
            MyBase.New(settings, sampleRate)
        End Sub

        Public Sub ShiftPitch(ByVal inputBuff() As Single, ByVal inputPitch As Single,
                              ByVal targetPitch As Single, ByVal outputBuff() As Single, ByVal nFrames As Integer)
            UpdateSettings()
            detectedPitch = inputPitch
            Dim shiftFactor = 1.0F
            If Me.settings.SnapMode Then
                If inputPitch > 0 Then
                    shiftFactor = snapFactor(inputPitch)
                    shiftFactor += addVibrato(nFrames)
                End If
            Else
                Dim midiPitch = targetPitch
                shiftFactor = 1.0F
                If inputPitch > 0 AndAlso midiPitch > 0 Then
                    shiftFactor = midiPitch / inputPitch
                End If
            End If

            If shiftFactor > 2.0 Then
                shiftFactor = 2.0F
            End If
            If shiftFactor < 0.5 Then
                shiftFactor = 0.5F
            End If

            ' fftFrameSize was nFrames but can't guarantee it is a power of 2
            ' 2048 works, let's try 1024
            Dim fftFrameSize = 2048
            Dim osamp = 8 ' 32 is best quality
            SmbPitchShift.smbPitchShift(shiftFactor, nFrames, fftFrameSize, osamp, Me.sampleRate, inputBuff, outputBuff)

            'vibrato
            'addVibrato(outputBuff, nFrames);

            shiftedPitch = inputPitch * shiftFactor
            updateShifts(detectedPitch, shiftedPitch, Me.currPitch)
        End Sub

        Private Sub UpdateSettings()
            'these are going here, because this gets called once per frame
            vibRate = Me.settings.VibratoRate
            vibDepth = Me.settings.VibratoDepth
            attack = CInt(Fix((Me.settings.AttackTimeMilliseconds * 441) / 1024.0))
        End Sub
    End Class


    Friend Class PitchShift
        Public Sub New(ByVal detected As Single, ByVal shifted As Single, ByVal destNote As Integer)
            Me.DetectedPitch = detected
            Me.ShiftedPitch = shifted
            Me.DestinationNote = destNote
        End Sub

        Private _detectedPitch As Single
        Public Property DetectedPitch As Single
            Get
                Return _detectedPitch
            End Get
            Private Set(ByVal value As Single)
                _detectedPitch = value
            End Set
        End Property

        Private _shiftedPitch As Single
        Public Property ShiftedPitch As Single
            Get
                Return _shiftedPitch
            End Get
            Private Set(ByVal value As Single)
                _shiftedPitch = value
            End Set
        End Property

        Private _destinationNote As Integer
        Public Property DestinationNote As Integer
            Get
                Return _destinationNote
            End Get
            Private Set(ByVal value As Integer)
                _destinationNote = value
            End Set
        End Property

        Public Overrides Function ToString() As String
            Return String.Format("detected {0:f2}Hz, shifted to {1:f2}Hz, {2}{3} ",
                                 DetectedPitch, ShiftedPitch, CType(DestinationNote Mod 12, Note), DestinationNote \ 12)
        End Function
    End Class
End Namespace
